pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -476,6 +478,38 @@ impl WalreceiverState {
         None
     }
 
+    fn select_connection_candidate(
```

Comment:
- This function deserves a comment.
- I initially doubted that considering only smallest_attempts_count is good. It bears a risk of choosing lagged node. The problem we are trying to solve is avoiding reconnecting to node which just gone dead. Another possible way would be immediately after getting connection failure to delete `wal_stream_candidates` entry; dead node won't send updates anymore. It has its own faults though; an obvious race -- we might still get update which was just slow to arrive -- and it doesn't consider split situations where etcd is visible but direct connection fails. The middle between these options can be augment this PR with forgetting `wal_connection_attempts` after period of time with no failures. But I'm ok merging as is, it is good enough currently.
- Nitpicking, but perhaps in more functional style (fold wal_connection_attempts to find min, then filter candidates) it would be simpler.
- s/assempts/attempts

Reply:
> I initially doubted that considering only smallest_attempts_count is good.
I've decided to leave the candidates selection logic as is, without removals or expirations, just since it's simpler.
I plan to revisit the global task restarts here at some point, can continue with this one too unless any problems pop up before.

Otherwise reworked.